Add READ_ALL_COMMANDS

markuspf opened the pull request:
This function allows to interpret all statements from a
InputTextStream until EOF is reached.

The result is either `fail` or a list of lists of up to two entries, one for
each statement:
 * If the stream cannot be opened the result is `fail`
 * If the execution of the statement was successful, then the
   first entry of the list is true. If the statement had a value
   then the second entry is this value
 * If the execution of a statement failed, then the first entry
   of the list is false.

markuspf commented:
I am using this to interpret Jupyter-Cells with more than one statement at the moment.

stevelinton commented:
One would assume that it would be possible to have a function that read one command, returned status and results and left the stream ready to read the next command. If so it would make sense to implemement this function as a GAP level loop over that one. I think it actually isn't -- the scanner state and the state of the stream aren't cleanly separable. I'm just adding this comment to prevent other people wasting time looking into the same thing.

markuspf commented:
I attempted making a function that reads just one command from a stream, and I have spent (wasted?) quite some time on it: The main issue is that the scanner reads in as much of the stream as it can (so, usually all of it), and then closes it afterwards. I have some ongoing work to address this, but since it is far bigger surgery than this patch, I decided to first do this here to get the Jupyter kernel working better.

fingolfin commented:
Would it be possible to add a simple test for the new function? It could just read a few simple commands from a string stream.

markuspf commented:
I added some tests and changed the function a bit to use `ReadEvalCommand` directly, and to handle errors better. (Which of course are things I noticed while adding the tests).

markuspf commented:
I think I might know how to implement `READ_ONE_COMMAND` now, but lets not hold back on this function, because it makes it possible to push the Jupyter kernel forward.

fingolfin commented on src/streams.c:
```diff
@@ -98,6 +98,67 @@ Int READ_COMMAND ( void )
+    Int result_count, result_capacity;
+    Obj result, result_list;
+
+    // 16 Results should be enough for everyone
```
that comment scared me, but it turns out that later on, you actually grow the list if necessary... I guess the comment is meant to be snarky ("640kb is enough for everyone" ?) but perhaps here I'd avoid that.

fingolfin commented on src/streams.c:
```diff
@@ -98,6 +98,67 @@ Int READ_COMMAND ( void )
+
+    /* try to open the file                                                */
+    if (!OpenInputStream(stream) ) {
+        return result_list;
```
So in case of failure, you return `[[false]]` ? What is the supposed output of your function? I would appreciate a brief comment before the function which summarizes what it does, and what its output is. 

For now, my guess is that it's a a list of "output lists" similar to what e.g. `CALL_WITH_CATCH` return, i.e. either `[false]` if there was an error, or else `[true]` or `[true,result]`.

Assuming that guess is right, doesn't that mean that I cannot distinguish failing to open a stream from a stream with a single command that produced an error? Wouldn't it then be better to return `Fail` here?

fingolfin replied:
Looking at your test case, I am surprised to see that an empty input also yields `[[false]]`, I would have expected an empty list... Perhaps you have good reasons for all this behavior, but then it should be (at least briefly) explained in source comments.

Otherwise, I'd start the function roughly like this in order to achieve the behavior I suggested:
```C
  if (!OpenInputStream(stream) )
    return Fail;

  result_list = NEW_PLIST( T_PLIST, result_capacity );
  SET_LEN__PLIST(result_list, 0);

  TLS(Input)->echo = (echo == True);

   do {
    ...
   } while(...);
```

fingolfin commented on src/streams.c:
```diff
@@ -98,6 +98,67 @@ Int READ_COMMAND ( void )
+{
+    ExecStatus status;
+    Int result_count, result_capacity;
+    Obj result, result_list;
```
Minor nitpick: Most of the GAP kernel seems to use camelCase, to I find it a bit jarring to see snake_case here.

markuspf commented:
This will need some history rewrite I guess.

fingolfin commented on src/streams.c:
```diff
@@ -98,6 +98,59 @@ Int READ_COMMAND ( void )
+{
+    ExecStatus Status;
+    Int ResultCount, ResultCapacity;
+    Obj Result, ResultList;
```
ahh, please don't hit me, but: we also tend to start variable names in lower case, uppercase is for functions... :)

fingolfin commented on src/streams.c:
```diff
@@ -98,6 +98,59 @@ Int READ_COMMAND ( void )
     return 1;
 }
 
+Obj FuncREAD_ALL_COMMANDS( Obj self, Obj stream, Obj echo )
```
Could you insert at least a minimal doc comment? Just 1-2 lines. And perhaps mention that this is used by the Jupyter GAP interface. Otherwise, a future code cleanup might decide to discard this function, as nothing uses it... ;-)..
